### PR TITLE
Export `ProjectManager` and `ProjectExplorer` APIs

### PR DESCRIPTION
**Pull request by SanjulaGanepola**
### Changes
- Export `ProjectManager` and `ProjectExplorer`
- Add more APIs to retrieve projects in `ProjectManager`
- Add API to push extensible children
- Add abstract `ProjectExplorerTreeItem`
- Refactor existing tree items to use the new abstract class

### CODE REVIEW

File: src/views/projectExplorer/library.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  async getChildren(): Promise<ProjectExplorerTreeItem[]> {
+    let items: ProjectExplorerTreeItem[] = [];
+
+    const ibmi = getInstance();
+    const files = await ibmi?.getContent().getObjectList({
+      library: this.name
+    });
+    if (files) {
+      for (const file of files) {
+        const path = `/QSYS.LIB/${this.name}/${file.name}`;
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/library.ts:34]: consider that we will have to support iASP
/<iasp/QSYS.LIB/....
@worksofliam is there some code that Sanjula should be modelling off of

Comment (edmundreinhardt) [reply src/views/projectExplorer/library.ts:34]: You probably will have a shared function to build this path and put it in the `ibmi` object?

File: src/views/projectExplorer/objectlibrary.ts [hunk 2]
```diff
@@ -2,17 +2,53 @@
...
 
-export default class ObjectLibrary extends TreeItem {
-  static contextValue = `objectLibrary`;
   constructor(public workspaceFolder: WorkspaceFolder) {
     super("Object Libraries", TreeItemCollapsibleState.Collapsed);
 
     this.resourceUri = workspaceFolder.uri;
     this.iconPath = new ThemeIcon(`root-folder`);
-    this.contextValue = `objectLibrary`;
+    this.contextValue = this.contextValue;
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/objectlibrary.ts:19]: Again this seems like a noop

File: src/views/projectExplorer/objectlibrary.ts [hunk 3]
```diff
@@ -2,17 +2,53 @@
  * (c) Copyright IBM Corp. 2023
  */
 
-import { ThemeIcon, TreeItem, TreeItemCollapsibleState, WorkspaceFolder } from "vscode";
+import { ThemeIcon, TreeItemCollapsibleState, WorkspaceFolder } from "vscode";
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/objectlibrary.ts:5]: Can we have a line or 2 of comments
I am not clear if this is representing an individual library or the library list

Comment (edmundreinhardt) [reply src/views/projectExplorer/objectlibrary.ts:5]: After reading more code I see that this is the heading for all of the target object libraries.

File: src/views/projectExplorer/variables.ts [hunk 4]
```diff
@@ -2,15 +2,44 @@
...
+    const possibleVariables = iProject?.getVariables();
+    const actualValues = await iProject?.getEnv();
+
+    if (possibleVariables && actualValues) {
+      items.push(...possibleVariables?.map(
+        varName => {
+          return new Variable(this.workspaceFolder, varName, actualValues[varName]);
+        }
+      ));
+
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/variables.ts:36]: So if there is a variable with no value in the env, how do we represent that variable

**Review by edmundreinhardt — CHANGES_REQUESTED** [review]

File: src/extension.ts [hunk 5]
```diff
@@ -8,10 +8,12 @@ import { loadBase, getInstance } from './ibmi';
 import { ProjectManager } from './projectManager';
 import JobLog from './views/jobLog';
 import ProjectExplorer from './views/projectExplorer';
+import { ProjectExplorerApi } from './projectExplorerApi';
+import { ProjectExplorerTreeItem } from './views/projectExplorer/projectExplorerTreeItem';
 
 // this method is called when your extension is activated
 // your extension is activated the very first time the command is executed
-export function activate(context: vscode.ExtensionContext) {
+export function activate(context: vscode.ExtensionContext): ProjectExplorerApi {
```
[the remark this answers is not in the document]
Comment (SanjulaGanepola) [reply src/extension.ts:16]: We should only be doing it on iproj.json, .ibmi.json. and .env file changes. I have updated the watcher's GlobPattern.

File: src/views/projectExplorer/library.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { ContextValue } from "../../projectExplorerApi";
+
+export default class Library extends ProjectExplorerTreeItem {
+  static contextValue = ContextValue.library;
+  name: string;
+
+  constructor(public workspaceFolder: WorkspaceFolder, path: string, name: string) {
+    super(name, TreeItemCollapsibleState.Collapsed);
+    this.description = path;
+    this.contextValue = this.contextValue;
```
[the remark this answers is not in the document]
Comment (SanjulaGanepola) [reply src/views/projectExplorer/library.ts]: This is actually a bug. Each tree item class has a statically defined `contextValue` which is what we should be assigning as the tree item's `contextValue`. Pushed a fix for this.

File: src/views/projectExplorer/library.ts (hunk 1 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/library.ts:37]: Planning to add iASP support in a separate PR. Issue to track this: #6

File: src/views/projectExplorer/objectlibrary.ts (hunk 2 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/objectlibrary.ts]: Fixed, thanks for catching that

File: src/views/projectExplorer/objectlibrary.ts (hunk 3 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/objectlibrary.ts]: Added a comment for each tree item

File: src/views/projectExplorer/variables.ts (hunk 4 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/variables.ts]: If there is a variable with no value, the description will indicate this. There is also a red decoration on the tree item to make it clear.

image

File: src/extension.ts [hunk 7]
```diff
@@ -125,10 +127,12 @@ export function activate(context: vscode.ExtensionContext) {
 				createFilePreview(content, library, file, member);
 			} catch (e) {
 				// if this works its a source member
-				vscode.commands.executeCommand('vscode.open', memberUri)
+				vscode.commands.executeCommand('vscode.open', memberUri);
 			}
 		}
 	});
+
+	return { projectManager: ProjectManager, projectExplorer: projectExplorer };
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/extension.ts:135]: Let's aim for the `types` as the folder which contains the types.

File: src/projectManager.ts [hunk 8]
```diff
@@ -68,15 +70,42 @@ export class ProjectManager {
...
+        }
+    }
+
     public static getProjectFromUri(uri: Uri): IProject | undefined {
-        const workspaceFolders = workspace.workspaceFolders;
-        if (workspaceFolders && workspaceFolders.length > 0) {
-            const changedWorkspaceFolder = workspaceFolders.filter(workspaceFolder =>
-                uri.fsPath.startsWith(workspaceFolder.uri.fsPath)
-            )[0];
+        for (const index in this.loaded) {
```
Comment (worksofliam) [inline src/projectManager.ts:92]: This for loop can be replaced by a simple `workspace.getWorkspaceFolder` API call, which returns `WorkspaceFolder`, which has the project index on it.

File: src/views/projectExplorer/project.ts [hunk 9]
```diff
@@ -14,4 +29,68 @@ export default class Project extends TreeItem {
...
+          arguments: [this.workspaceFolder],
+          title: `Create project .env`
+        }
+      }));
+    }
+
+    items.push(new ObjectLibraries(this.workspaceFolder));
+
+    for await (const extensibleChildren of Project.callBack) {
+      this.extensibleChildren.push(...await extensibleChildren(iProject!));
```
Comment (worksofliam) [inline src/views/projectExplorer/project.ts:86]: Something to consider is how we are going to handle errors from the `extensibleChildren` callback. If `extensibleChildren` crashes here, then nothing will be returned and the view will be blank.

**Review by worksofliam — CHANGES_REQUESTED** [review]
This is a really great cleanup. The introduction of `getChildren` is super awesome.

Take a look at my comments and reassign to me whenever you're ready for another review.

File: src/projectManager.ts (hunk 8 above)
Comment (SanjulaGanepola) [reply src/projectManager.ts]: Good catch. Updated this

File: src/views/projectExplorer/project.ts (hunk 9 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/project.ts]: Added a try catch around the callback

**Comment by SanjulaGanepola** [comment]
Since this PR already has a lot of changes, I think it would be best to merge it and then work on adding the `types` folder in a separate PR.

**Review by edmundreinhardt — APPROVED** [review]
Looks good

**Review by edmundreinhardt — APPROVED** [review]
Looks good

**Pull request merged**